test/e2e/gameserver_test.go

```diff
@@ -213,6 +213,113 @@ func TestGameServerUnhealthyAfterDeletingPod(t *testing.T) {
+	assert.NoError(t, err)
+
+	logger.WithField("gs", newGs.ObjectMeta.Name).Info("GameServer created")
+
+	address := fmt.Sprintf("%s:%d", newGs.Status.Address, newGs.Status.Ports[0].Port)
+	logger.WithField("address", address).Info("Dialing UDP message to address")
+
+	logger.Info("crashing, and waiting to see restart")
+	err = wait.PollImmediate(3*time.Second, 3*time.Minute, func() (bool, error) {
+		pod, err := framework.KubeClient.CoreV1().Pods(defaultNs).Get(newGs.ObjectMeta.Name, metav1.GetOptions{})
+		if err != nil {
+			logger.WithError(err).Warn("could not get pod for gameserver")
+			return true, err
+		}
+
+		for _, c := range pod.Status.ContainerStatuses {
+			if c.Name == newGs.Spec.Container && c.RestartCount >= 1 {
+				logger.Info("successfully crashed. Moving on!")
+				return true, nil
+			}
+		}
+
+		// create a connection each time, as weird stuff happens if the receiver isn't up and running.
+		conn, err := net.Dial("udp", address)
+		assert.NoError(t, err)
+		defer conn.Close() // nolint: errcheck
+		// doing this last, so that there is a short delay between the message being sent, and the check.
+		// also gives the process a chance to come back up
+		_, err = conn.Write([]byte("CRASH"))
+		logger.WithError(err).Info("no crash, trying again")
+		return false, nil
+	})
+	assert.NoError(t, err)
+
+	// check that the GameServer is not in an unhealthy state. If it does happen, it should happen pretty quick
+	newGs, err = framework.WaitForGameServerState(newGs, agonesv1.GameServerStateUnhealthy, 5*time.Second)
+	// should be an error, as the state should not occur
+	if !assert.Error(t, err) {
+		assert.FailNow(t, "GameServer should not be Unhealthy")
+	}
+	assert.Contains(t, err.Error(), "waiting for GameServer")
+
+	// ping READY until it doesn't fail anymore - since it may take a while
+	// for this to come back up -- or we could get a delayed CRASH, so we have to
+	// wait for the process to restart again to fire the SDK.Ready()
+	logger.Info("marking GameServer as ready")
+	err = wait.PollImmediate(3*time.Second, 3*time.Minute, func() (bool, error) {
+		newGs, err = gsClient.Get(newGs.ObjectMeta.Name, metav1.GetOptions{})
+		if err != nil {
+			logger.WithError(err).Warn("could not get gameserver")
+			return true, err
+		}
+
+		if newGs.Status.State == agonesv1.GameServerStateReady {
+			logger.Info("ready! Moving On!")
+			return true, nil
+		}
+
+		// create a connection each time, as weird stuff happens if the receiver isn't up and running.
+		conn, err := net.Dial("udp", address)
+		assert.NoError(t, err)
+		defer conn.Close() // nolint: errcheck
+		// doing this last, so that there is a short delay between the message being sent, and the check.
+		_, err = conn.Write([]byte("READY"))
+		logger.WithError(err).WithField("gs", newGs.ObjectMeta.Name).WithField("state", newGs.Status.State).Info("no ready, trying again")
+		return false, nil
+	})
+	if !assert.NoError(t, err) {
```

Comment:
elsewhere I just see `assert.NoError()` but this one wraps that and calls `FailNow`. Is that to short circuit running the remainder of the test? If so, would it be simpler to just write

```go
if err != nil {
    assert.FailNow(t, ...)
}
```

I find the statements that check the boolean result of the assert to be hard to read, and they are basically just checking if err is or isn't nil and calling another assert, which can be done more simply.